### ISSUE TITLE
[NativeAOT] Missing delegate marshalling data for MMI

### ISSUE THREAD

**Issue by prajaybasu** [dotnet/runtimelab#1770]
```
Unhandled exception!
      System.TypeInitializationException: A type initializer threw an exception. To determine which type, inspect the InnerException's StackTrace property.
       ---> EETypeRva:0x00BAC1C8(System.Runtime.InteropServices.MissingInteropDataException): Microsoft.Management.Infrastructure.Native.DeserializerCallbacks+GetIncludedFileBufferAppDomainProxyDelegate is missing delegate marshalling data. To enable delegate marshalling data, add a MarshalDelegate directive to the application rd.xml file. For more information, please visit http://go.microsoft.com/fwlink/?LinkID=393965
         at Internal.Runtime.CompilerHelpers.RuntimeInteropData.GetDelegateMarshallingStub(RuntimeTypeHandle, Boolean) + 0x92
         at System.Runtime.InteropServices.PInvokeMarshal.AllocateThunk(Delegate) + 0x7a
         at System.Runtime.CompilerServices.ConditionalWeakTable`2.GetValueLocked(TKey, ConditionalWeakTable`2.CreateValueCallback) + 0x26
         at System.Runtime.CompilerServices.ConditionalWeakTable`2.GetValue(TKey, ConditionalWeakTable`2.CreateValueCallback) + 0x3e
         at System.Runtime.InteropServices.PInvokeMarshal.GetFunctionPointerForDelegate(Delegate) + 0x92
         at <Module>.Microsoft.Management.Infrastructure.Native.MI_ApplicationWrapper_Initialize_Managed(_MI_ApplicationWrapper*) + 0x66
```

There aren't many examples around on how to deal with MarshalDelegate and I can't really do much about MMI since it's closed source (at least the C++/CLI part). And WMI requires COM Interop. It would be really nice to have a working WMI alternative with NativeAOT support.

**Comment by MichalStrehovsky** [comment]
After #1772 is merged, this should help:

```xml
<Directives>
  <Application>
    <Assembly Name="XXXXXXX">
      <Type Name="Microsoft.Management.Infrastructure.Native.DeserializerCallbacks+GetIncludedFileBufferAppDomainProxyDelegate" MarshalDelegate="Required All" />
    </Assembly>
  </Application>
</Directives>
```

You need to replace the XXXXXXX with the name of the assembly that has the type. I can't tell from the message.

**Comment by prajaybasu** [comment]
Are the changes available in the dotnet-experimental feed? I'm on `7.0.0-alpha.1.21607.1` and I'm getting the same error.

>  assembly that has the type. I can't tell from the message.

It's `Microsoft.Management.Infrastructure.Native`.  It's closed source and seems to be written to work with C++/CLI, so I don't know if it's supposed to work in the first place.

https://github.com/PowerShell/MMI/blob/master/src/Microsoft.Management.Infrastructure/Native/MI_Deserializer.cs

**Comment by MichalStrehovsky** [comment]
21607.2 is the first that has the fix.

**Comment by prajaybasu** [comment]
Thanks for the fix, that worked.

I had to use this rd.xml file, for anyone interested


```
<Directives xmlns="http://schemas.microsoft.com/netfx/2013/01/metadata">
	<Application>
		<Assembly Name="Microsoft.Management.Infrastructure.Native">
			<Type Name="Microsoft.Management.Infrastructure.Native.DeserializerCallbacks+ClassObjectNeededAppDomainProxyDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.DeserializerCallbacks+ClassObjectNeededCallbackDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.DeserializerCallbacks+GetIncludedFileBufferAppDomainProxyDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.DeserializerCallbacks+GetIncludedFileBufferCallbackDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.SessionHandleCallbackDefinitions+SessionHandle_ReleaseHandle_CallbackWrapper_Release_Managed_Delegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.SessionHandleCallbackDefinitions+SessionHandle_ReleaseHandle_CallbackWrapper_Invoke_Managed_Delegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.OperationCallbacksDefinitions+InstanceResultAppDomainProxyDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.OperationCallbacksDefinitions+StreamedParameterResultAppDomainProxyDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.OperationCallbacksDefinitions+WriteMessageAppDomainProxyDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.OperationCallbacksDefinitions+WriteProgressAppDomainProxyDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.OperationCallbacksDefinitions+WriteErrorAppDomainProxyDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.OperationCallbacksDefinitions+PromptUserAppDomainProxyDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.OperationCallbacksDefinitions+ClassAppDomainProxyDelegate" MarshalDelegate="Required All" />
			<Type Name="Microsoft.Management.Infrastructure.Native.OperationCallbacksDefinitions+IndicationAppDomainProxyDelegate" MarshalDelegate="Required All" />
			<Type Name="GCHandleHelper+FreeGcHandleDelegate" MarshalDelegate="Required All" />	
		</Assembly>
	</Application>
</Directives>
```

I also generated a `mi.lib` file for Windows from mi.dll added this to my .csproj

```
	<ItemGroup>
		<RdXmlFile Include="rd.xml" />
		<DirectPInvoke Include="mi" />
		<NativeLibrary Include="mi.lib" Condition="$(RuntimeIdentifier.StartsWith('win'))" />
	</ItemGroup>
```
